tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs

```diff
@@ -89,5 +94,208 @@ public void Initialize()
+
+            // Call acquire token
+            var result = AssertException.TaskThrows<AdalException>(() =>
+            context.AcquireTokenAsync(TestConstants.DefaultResource, TestConstants.DefaultClientId,
+                userPasswordCredential));
+
+            // Check inner exception
+            Assert.AreEqual(" Response status code does not indicate success: 404 (NotFound).", result.InnerException.Message);
+
+            // All mocks are consumed
+            Assert.AreEqual(0, HttpMessageHandlerFactory.MockHandlersCount());
+        }
+
+        [TestMethod]
+        [Description("Test for AcquireToken with valid token in cache")]
+        public async Task AcquireTokenWithValidTokenInCache_ReturnsCachedTokenTestAsync()
+        {
+            AuthenticationContext context = new AuthenticationContext(TestConstants.DefaultAuthorityCommonTenant, new TokenCache());
+            await context.Authenticator.UpdateFromTemplateAsync(null);
+
+            TokenCacheKey key = new TokenCacheKey(TestConstants.DefaultAuthorityCommonTenant,
+                TestConstants.DefaultResource, TestConstants.DefaultClientId, TokenSubjectType.User,
+                TestConstants.DefaultUniqueId, TestConstants.DefaultDisplayableId);
+            context.TokenCache.tokenCacheDictionary[key] = new AuthenticationResultEx
+            {
+                RefreshToken = "some-rt",
+                ResourceInResponse = TestConstants.DefaultResource,
+                Result = new AuthenticationResult("Bearer", "existing-access-token",
+                    DateTimeOffset.UtcNow + TimeSpan.FromMinutes(100))
+            };
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler("https://login.microsoftonline.com/common/oauth2/token")
+            {
+                Method = HttpMethod.Post,
+                ResponseMessage = MockHelpers.CreateSuccessTokenResponseMessage(),
+                PostData = new Dictionary<string, string>()
+                {
+                    {"resource", TestConstants.DefaultResource},
+                    {"client_id", TestConstants.DefaultClientId},
+                    {"grant_type", "refresh_token"},
+                    {"scope", "openid"}
+                }
+            });
+
+            // Call acquire token
+            AuthenticationResult result = await context.AcquireTokenAsync(TestConstants.DefaultResource, TestConstants.DefaultClientId,
+                new UserPasswordCredential(TestConstants.DefaultDisplayableId, TestConstants.DefaultPassword));
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual("existing-access-token", result.AccessToken);
+
+            // There should be one cached entry.
+            Assert.AreEqual(1, context.TokenCache.Count);
+            Assert.IsNotNull(result.UserInfo);
+        }
+
+        [TestMethod]
+        [Description("Mex endpoint fails to resolve and results in a 404")]
+        public async Task MexEndpointFailsToResolveTestAsync()
```

Comment:
how is this test different than the one you wrote above for mex failure? you should repurpose this test to return a malformed XML and make sure XML parser fails.

Reply:
This makes more sense...thanks